diff -up evolution-3.28.5/src/addressbook/gui/widgets/eab-contact-merging.c.contacts-prefer-orig-value evolution-3.28.5/src/addressbook/gui/widgets/eab-contact-merging.c
--- evolution-3.28.5/src/addressbook/gui/widgets/eab-contact-merging.c.contacts-prefer-orig-value	2018-07-30 15:37:05.000000000 +0200
+++ evolution-3.28.5/src/addressbook/gui/widgets/eab-contact-merging.c	2022-09-15 08:19:23.228037822 +0200
@@ -512,7 +512,12 @@ mergeit (EContactMergingLookup *lookup)
 					G_CALLBACK (dropdown_changed), data);
 				g_object_set_data_full (G_OBJECT (dropdown), "eab-contact-merging::dropdown-data", data, g_free);
 
-				if (field == E_CONTACT_NICKNAME || field == E_CONTACT_GIVEN_NAME || field == E_CONTACT_FAMILY_NAME || field == E_CONTACT_FULL_NAME)
+				/* Only prefer the original value when it's filled */
+				if (string1 && *string1 && (
+				    field == E_CONTACT_NICKNAME ||
+				    field == E_CONTACT_GIVEN_NAME ||
+				    field == E_CONTACT_FAMILY_NAME ||
+				    field == E_CONTACT_FULL_NAME))
 					gtk_combo_box_set_active (GTK_COMBO_BOX (dropdown), 1);
 				else
 					gtk_combo_box_set_active (GTK_COMBO_BOX (dropdown), 0);
